Bug 1315808 - New Log Viewer integration

helfi92 opened the pull request:
Main Features
* Be able to highlight different error lines
* Get rid of "double hashing" when linking to a specific line number
* Smooth scrolling
* Highlight single and multiple lines
* Follow log (option to scroll to the bottom of the log)

<!-- Reviewable:start -->

---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/mozilla/treeherder/1975)
<!-- Reviewable:end -->

autolander commented:
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

helfi92 commented:
I only commented the lines which were not making the code pass the linting tests. Most of the old implementation is still uncommented, but simply not used.

wlach commented:
@helfi92 it looks like the ui tests are failing, could you have a look? http://treeherder.readthedocs.io/ui/installation.html#running-the-unit-tests

Also feel free to just delete any old code that is unused. I'd rather not have it sitting around in the repository, as it's going to confuse people.

helfi92 commented:
@wlach Sounds good. I will delete any old code that is unused and then push a commit. By the way, `grunt checkjs` and `./tests/ui/scripts/test.sh` both execute successfully. But when I push my commit, I get a `Firefox 31.0.0 (Linux 0.0.0) ERROR`. Does that mean I will need to download a Firefox version 31 in order to fix that issue locally?

wlach commented:
@helfi92 you shouldn't, no. hmm. could you rebase on master and reinstall the node dependencies? @jgraham updated the karma firefox dep (which we use to start the browser) in b379a107590e29c123a354eefae463d29580c7aa

helfi92 commented:
@wlach I did merge master with my branch in both of my commits and installed the new dependencies, so that's probably not the problem. I have a feeling that it might be happening because of me using ES6 features. Should I just change my code to be using ES5?

wlach commented:
@helfi92 In theory es6 should be fine, we're already using features from it in various places. If you could try to figure it out that would be good, if you've tried and are still stuck, let me know and I can take a peek.

eliperelman commented on ui/js/controllers/logviewer.js:
```diff
@@ -376,6 +401,21 @@ logViewerApp.controller('LogviewerCtrl', [
 
+        function displayErrorLines(errors) {
+            var css = errors
+              .reduce((line, err) => `${line}a[id="${err.line_number + 1}"]+span, `, '')
```
What does `line` mean here?

helfi92 replied:
@eliperelman `line` contains the selector (e.g, `a[id="1829"]+span, a[id="4490"]+span, a[id="4492"]+span, a[id="4538"]+span`). Want me to rename it to `selector`

eliperelman replied:
Combined with my other comment then:

```js
errors
    .map({ line_number } => `a[id="${line_number + 1}"] + span`)
    .join(', ')
```

helfi92 replied:
@eliperelman would it be smart to remove all the spaces to make the string smaller? No spaces between the `+` sign and after the comma `,`

```javascript
var css = errors
  .map(({ line_number }) => `a[id="${line_number + 1}"]+span`)
  .join(',')
```

eliperelman commented on ui/js/controllers/logviewer.js:
```diff
@@ -376,6 +401,21 @@ logViewerApp.controller('LogviewerCtrl', [
+        function displayErrorLines(errors) {
+            var css = errors
+              .reduce((line, err) => `${line}a[id="${err.line_number + 1}"]+span, `, '')
+              .slice(0, -2)
```
What does the `.slice` here do?

helfi92 replied:
@eliperelman it slices the last space character with the last comma. But now that you pointed this out, i can simply not add a space between each selector and simply slice the last comma before doing a `concat` with the style object. Thoughts?

eliperelman replied:
Ah I see now. How about this?

```js
errors
    .map((line, err) => `${line}a[id="${err.line_number + 1}"] + span`)
    .join(', ')
```

helfi92 commented on ui/js/directives/treeherder/log_viewer.js:
```diff
@@ -0,0 +1,30 @@
+            var searchPart = () => {
+                var q = $location.search();
+
+                return "&highlightStart=" + q.highlightStart + "&highlightEnd=" + q.highlightEnd + "&lineNumber=" +
```
@eliperelman Yes we can, but the literal will have a couple of backslashes since template literals records the newline character.

helfi92 replied:
I like the array-building proposition. I'll use that one. Thanks!

eliperelman commented:
So, I there may be one issue here which is live logs. Right now the unified logviewer gives preference to non-live logs by attempting to load them as a binary arraybuffer first, and then gives up after 10 seconds and tries again as plain text. If we can know in Treeherder when loading the log whether it is still live or not, we can add a flag to the Unified Logviewer to skip loading as array buffer and go straight to text. Thoughts @helfi92 @wlach?

wlach commented:
@helfi92 Everything passes here after rebasing on master. Could you push a branch rebased on master and see if that fixes travis?

P.S. To test locally you need to install a version of Firefox. It'll just use whatever version you have installed.

helfi92 commented:
@eliperelman from my understanding, the current implementation of treeherder does not allow you to view live logs, only completed ones.

helfi92 commented:
@wlach I rebased on master but I am still getting the same errors. It doesn't show any errors on my machine though, just on travis
<img width="885" alt="screen shot 2016-11-08 at 4 36 32 pm" src="https://cloud.githubusercontent.com/assets/3766511/20118189/9fe64d18-a5d1-11e6-9248-16e213735ce5.png">

eliperelman commented on ui/js/controllers/logviewer.js:
```diff
@@ -376,6 +401,21 @@ logViewerApp.controller('LogviewerCtrl', [
+            var css = errors
+              .map(({ line_number }) => `a[id="${line_number + 1}"]+span`)
+              .join(',')
+              .concat('{background-color:red;color:white}');
```
I also think that `background-color: pink; color: red;` has a more appealing look (for an error).

wlach commented:
@helfi92 It just occured to me that the version of Firefox in the travis environment is rather old (31). Maybe there's an es6 feature it doesn't properly support or so.

Do you want to try upgrading it to the latest and see if it helps? Should be a minor tweak to .travis.yml in the root directory. https://docs.travis-ci.com/user/firefox/

helfi92 commented:
@wlach That seems to be the fix. the ui-tests are now passing :) Thanks for taking a look and helping. I will go ahead now and delete the unused code from the old logviewer then it should be ready for review

helfi92 commented:
@wlach I removed the unused code from the old log viewer. Code is now ready to reviewed

wlach commented:
Haven't yet looked at the code in detail but a few problems I'm noticing in testing:

1. The highlighting looks wrong on my Linux nightly:
![Uploading Screenshot from 2016-11-18 10-51-58.png…]()

2. Highlighting doesn't work at all if you enable line wrapping (which we probably want to enable by default, by the way)

3. It looks like the logviewer urls are now ridiculously long by default and expose a lot of internal state:

http://localhost:8000/logviewer.html#?job_id=39449331&repo=mozilla-inbound&highlightStart=&highlightEnd=&lineNumber=&wrapLines=false&showLineNumbers=&jumpToHighlight=false&followLog=false

These links are frequently shared so we should try to shorten them if at all possible. The default link  should be something like `http://localhost:8000/logviewer.html#?job_id=39449331&repo=mozilla-inbound`. I think the only really relevant piece of extra state that we might also want to pass in is a directive to jump to a specific line number.

4. I don't think the "follow log" option is relevant by default for treeherder, could we configure it out somehow?

Overall though, this looks like a great step forward. Thanks for doing this awesome work!

helfi92 commented:
@wlach Thanks for the kind words. @eliperelman is the man behind this awesome logviewer. With regards to your comment, highlighting actually works when you enable line wrapping, however the line jumping is off since each line can now have a different height. We were thinking of perhaps removing this functionality for now at least until we figure out a solution to it. I also agree with you, the logviewer URL is long and can definitely be shortened to be using line number only for treeherder purposes. 

Regarding the highlighting of errors, what is it that makes it look wrong. I have made it have a pink background with red text color. Let me know what you think is looking wrong so I can make it better :). For your last point, would removing `followLog` from the query URL answer your concern?

wlach commented:
@helfi92 It just looks wrong -- it looks like only half the line is highlighted (vertically anyways). Pink on red is also hard to read IMO, but that's another story. :)

I'd be fine if we removed the option to wrap the line for now (we don't have that feature anyway), though it would be really nice to have fixed at some point as IMO unwrapped lines can be much harder to read.

I would prefer if followLog were removed from both the URL and the user interface.

eliperelman commented:
Line wrapping and follow log can just be turned off by default, but I won't be removing follow log from the UI. You can pass in CSS to hide that feature along with the CSS for the error highlighting, and that should achieve the desired effect.

I agree the spacing of the error highlighting looks jacked. I remember recommending the pink/red scheme originally and it looked good, so maybe it will look better when the spacing is fixed.

I really hate the line wrapping feature, and fixing it would increase the complexity by an order of magnitude. We can either rip it out or keep it off and hide it in the UI with the CSS passing.

helfi92 commented:
@wlach made the changes you requested:

1- Make logviewer URL have the following format:
* http://localhost:8000/logviewer.html#?job_id=31592759&repo=try&lineNumber=1741
* http://localhost:8000/logviewer.html#?job_id=31592759&repo=try&lineNumber=1741-1782

2- Remove Follow Log (toolbar) from both the URL and the user interface.

3- This [PR](https://github.com/taskcluster/unified-logviewer/pull/22) should hopefully fix the error highlighting pointed out earlier.

Let me know if there is anything else you want me to add/modify/remove :)

wlach commented:
Thanks for the followup! Still a few issues.

Error highlighting still doesn't look right, it should cover the whole line:

![screenshot from 2016-11-21 15-12-10](https://cloud.githubusercontent.com/assets/20569/20498886/ed4e00a2-affc-11e6-8114-30494fc13aa1.png)

How difficult would it be to just swap the CSS you're using for what we have in the existing treeherder logviewer? There's a lot of things about it that I are rather nice (smaller font so you can see more, consistency with the rest of the UI, slightly more optimal spacing):

![screenshot from 2016-11-21 15-20-35](https://cloud.githubusercontent.com/assets/20569/20499158/28944792-affe-11e6-8f7d-4ed2a9b7c9de.png)

On my machine, it seems like scrolling the log doesn't work as I'd expect -- there's a pause between me making the scroll action and seeing the result. Would it be possible to pre-cache the entire log for smaller/typical files (<5M or so?).

helfi92 commented:
Hey @wlach, error highlighting still doesn't work because the following [PR](https://github.com/taskcluster/unified-logviewer/pull/22) has not yet been merged. The new log runs pretty fast on my machine so I'm not too sure why it's running slow on yours. @eliperelman are you also getting the same behavior? With regards to swapping the CSS i.e., smaller font and more optimal spacing, we would need to make some changes to the logviewer code because for now, the line jumping feature only works for lines with height `19px` Any thoughts on that @eliperelman

eliperelman commented:
These things can all be changed, but there needs to be an API for it, because those should not be in the agnostic logviewer. Color changes, padding, font sizes, etc. need to be given programmatically since the default theme of the logviewer can be used by anyone, not just treeherder.

The line height being modified is not currently possible; it would need a PR to the logviewer. If you would like me to check out a potentially slow-running log, please provide a link and I'll take a look.

wlach commented:
Ok, I am not sure if we should land this before we get the styling sync'ed up with what's currently there, as we might have user complaints due to the smaller viewing area.

The slowness I notice when scrolling a large amount on larger logs. E.g.:

http://localhost:8000/logviewer.html#?job_id=39655709&repo=mozilla-inbound&lineNumber=6063

Try jumping from end of the document to the other using the scrollbar.

I also found another bug. Navigating directly to this link doesn't seem to work (I get a console error):

http://localhost:8000/logviewer.html#?job_id=39655709&repo=mozilla-inbound

helfi92 commented:
@wlach I made the requested changes. This is how it now looks on my end.
<img width="1440" alt="screen shot 2016-11-28 at 3 38 42 pm" src="https://cloud.githubusercontent.com/assets/3766511/20685105/fe523178-b580-11e6-9d2b-c041e9acdcd9.png">

Let me know if this was what you wanted and whether you want me to change anything else. @eliperelman, did you have the chance to check whether you were having a delay scrolling a large log, more specifically the URL @wlach provided? It runs pretty smoothly on my laptop (firefox + chrome).

wlach commented:
That looks pretty ok to me though I'm still uncertain how sheriffs and others will react. After discussing this a bit during last week's treeherder meeting, we decided that we'll try putting this branch on stage and getting some feedback from the sheriffs and others. Thanks for your work so far!

wlach commented:
@helfi92 Actually, would you mind putting the branch up on a github pages repository for people to test? This will make it easier to iterate on updates if that proves to be necessary, as well as leave stage open for general testing. 

The procedure for doing this is described here:

http://treeherder.readthedocs.io/common_tasks.html#sharing-ui-only-changes-with-others-using-github-pages

Once that's done, we can write a message to a few newsgroups to try to encourage people to test it out. Sorry for not thinking of this earlier.

helfi92 commented:
@wlach Thanks for providing the procedures. Here's the link [https://helfi92.github.io/treeherder/ui/](https://helfi92.github.io/treeherder/ui/).

Example: [https://helfi92.github.io/treeherder/ui/logviewer.html#?job_id=40233990&repo=mozilla-inbound&lineNumber=24831](https://helfi92.github.io/treeherder/ui/logviewer.html#?job_id=40233990&repo=mozilla-inbound&lineNumber=24831)

helfi92 commented:
@wlach, what do you think about the new changes. I added some css to make it look like the current treeherder logviewer. Changes were pushed to `gh-pages` branch for quick access :)

helfi92 commented:
You might notice that there is a delay before the css changes are shown. That's because I am only sending it when the log viewer steps (top right view) is loaded. I will be working on making this process faster.

wlach commented:
@helfi92 this looks great! it seems more responsive now as well.

It seems like the hover effect for highlighting the error line quicklinks got broken though? Compare hovering over them here with your example: https://treeherder.mozilla.org/logviewer.html#?job_id=40233990&repo=mozilla-inbound&lineNumber=25195

I feel like functionally this is good enough to merge. I still need to review the code, will do so after the above is addressed.

helfi92 commented:
@wlach i addressed your comments regarding highlighting. Let me know if there's anything else you want me to change in the ui. otherwise, it should be ready for review.

wlach commented:
@helfi92 testing locally I'm still not seeing the hover effect on the error line links that I'm expecting. here's a screenshot of the current behaviour for clarity (note how line 28226 is highlighted)

<img width="647" alt="screen shot 2016-12-22 at 1 44 14 pm" src="https://cloud.githubusercontent.com/assets/20569/21436508/e5afe690-c84c-11e6-94e3-1257e9bcc649.png">

wlach commented:
(apologies if my previous comment was unclear)

wlach commented on ui/js/services/main.js:
```diff
@@ -29,7 +29,7 @@ treeherder.factory('thUrl', [
             },
             getLogViewerUrl: function(job_id) {
-                return "logviewer.html#?job_id=" + job_id + "&repo=" + $rootScope.repoName;
+                return "logviewer.html#?job_id=" + job_id + "&repo=" + $rootScope.repoName + "&lineNumber=";
```
Is it actually necessary to specify `&lineNumber=` here?

helfi92 replied:
you're right. i don't think it's necessary to specify that here. i'll remove it

wlach commented on ui/js/directives/treeherder/log_viewer.js:
```diff
@@ -0,0 +1,34 @@
+'use strict';
+
+treeherder.directive('thLogViewer', ['$sce', '$location', ($sce, $location) => {
```
Since we're here, we might want to switch to using an angular component instead of a directive, which is a bit simpler... there's a few examples in `ui/js/components/`

helfi92 commented:
@wlach Yes there was a misunderstanding. I thought you were referring to the log itself but now I see what you mean. Nevertheless, the log highlighting should now look exactly like the current implementation so that's at least good. I'll now work on the log viewer step highlight and change the directive to a component. Thanks for your comments :)

helfi92 commented:
@wlach I switched the directive into a component and I made sure highlighting works.

wlach commented:
@helfi92 Awesome! Looks great. Unfortunately I noticed one last feature that didn't make its way from the old logviewer implementation to the new. In the old one, it would auto-jump to the first failure line on load, this doesn't seem to be the case with the new one. E.g. compare when running a local instance:

http://localhost:8000/logviewer.html#?job_id=41099231&repo=mozilla-inbound
https://treeherder.mozilla.org/logviewer.html#?job_id=41099231&repo=mozilla-inbound

For reference, the original bug where this functionality was implemented was: https://bugzilla.mozilla.org/show_bug.cgi?id=1245760

After that's done, we can merge this into master for reals.

helfi92 commented:
@wlach My bad. I just realized when I made the transition from the directive to the component, the iframe's on load event wasn't firing at the right moment. I made the required change and it seems to be acting accordingly. 

By the way, I noticed when you open a log at first, it adds a certain offset to the lineNumber but then upon clicking different error lines from the log viewer steps , it removes that offset and so line jumping acts like the current implementation of this PR. Is that a wanted behaviour? I am assuming the added offset was used so that a user sees the lines right before the error.  Do you want me to add this feature or is this not needed anymore since there won't be any scrolling issues with the new logviewer.

wlach commented:
I'm pretty sure the old behavior was just an accident of implementation, let's just leave it the way it is for now and see if people complain. :)

I think this is good to merge now! Thank you so much for your work on this. There are some conflicts with what's currently on master, so I'll fix that up first.

wlach commented:
Created PR #2050 for actually merging this.
